adds support for including Azure VM metadata

jipperinbham opened the pull request:
### What does this PR do?

 When `collect_azure_tags` is set to true, specific instance metadata will be included as part of the host tags.

### Motivation

Similar functionality already exists for AWS and GCE. My company needs it for Azure as well. 

### Describe your test plan

Used the branch to build a binary and tested on internal infrastructure.

jipperinbham committed:
adds support for including Azure VM metadata

Similar functionality already exists for AWS and GCE. When collect_azure_tags is set to true, specific instance
metadata will be included as part of the host tags.

sarina-dd reviewed (approved):
👍 LGTM, docs

albertvaka commented:
Hi @jipperinbham! Thanks for implementing this! Note we are currently preparing a release and `master` is frozen, so we can't merge your patch now. Once the release is done, we can get back to this. Thanks again and sorry for the delay this introduces :)

jipperinbham commented:
Hi @albertvaka, I've now signed the CLA after review from myself and company. Please let me know if there's any changes needed!

hush-hush commented on pkg/util/azure/azure_tags.go:
```diff
@@ -0,0 +1,50 @@
+
+	metadataResponse, err := getResponse(metadataURL + "/metadata/instance/compute?api-version=2017-08-01")
+	if err != nil {
+		return tags, fmt.Errorf("unable to query metadata endpoint: %s", err)
```
Could we return `nil` instead of an empty slice for `tags` here ?

hush-hush commented on pkg/util/azure/azure_tags.go:
```diff
@@ -0,0 +1,50 @@
+
+	err = json.Unmarshal([]byte(metadataResponse), &metadata)
+	if err != nil {
+		return tags, err
```
Same here.

hush-hush commented on pkg/util/azure/test/azure_metadata.json:
```diff
@@ -0,0 +1,82 @@
+      }
+  },
+  "subscriptionId": "xxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx",
+  "tags": "Department:IT;Environment:Test;Role:WebRole",
```
We need to also pull the actual tags set by users.

jipperinbham replied:
I wasn't sure whether I should include this initially. Should there be an exclude filter similar to how the [GCE filter](https://github.com/DataDog/datadog-agent/blob/8086c3303dc785d560f1b73e031c4dc3a722ca0a/pkg/util/gce/gce_tags.go#L91-L100) works?

hush-hush replied:
Yes, might be a good idea

hush-hush commented on pkg/util/azure/azure_tags.go:
```diff
@@ -0,0 +1,50 @@
+}
+
+// GetTags gets the tags from the GCE api
+func GetTags() ([]string, error) {
```
You will need to check the `cloud_provider_metadata` field like here: https://github.com/DataDog/datadog-agent/blob/master/pkg/util/azure/azure.go#L37

hush-hush commented on releasenotes/notes/collect-azure-metadata-2773087e11627255.yaml:
````diff
@@ -0,0 +1,6 @@
+enhancements:
+  - |
+    Azure VM instance metadata may now be included as host
+    tags by using the ``collect_azure_tags``` configuration
````
```suggestion
    tags by using the `collect_azure_tags` configuration
```

hush-hush reviewed (changes requested):
Hi @jipperinbham,

Thanks for this PR. I added a few comments but this looks good overall.

graybjones commented on pkg/util/azure/azure_tags.go:
```diff
@@ -0,0 +1,50 @@
+
+	if metadata.VMID != "" {
+		tags = append(tags, fmt.Sprintf("vm-id:%s", metadata.VMID))
+	}
```
The crawler does not appear to tag the VM id. It does use it to determine the host name though. The VM id may show up as hostname if no other name is determined.

jipperinbham replied:
@graybjones I'm not familiar with the crawler you're referring to?

graybjones replied:
Sorry for the confusion. I will edit my comments for clarity.

jipperinbham replied:
I don't see a way to currently know whether or not the hostname has been set to determine whether or not to include this tag. Any suggestions?

graybjones commented on pkg/util/azure/azure_tags.go:
```diff
@@ -0,0 +1,50 @@
+	}
+	if metadata.VMSize != "" {
+		tags = append(tags, fmt.Sprintf("vm-size:%s", metadata.VMSize))
+	}
```
The crawler tags vm size as "size"

jipperinbham replied:
Sorry for just not getting back, I've verified we get tags for `resource_group` but we don't seem to be getting one for `size`. Is there a specific setting needed for that to be included in the default host tags?

graybjones commented on pkg/util/azure/azure_tags.go:
```diff
@@ -0,0 +1,50 @@
+	}
+	if metadata.ResourceGroup != "" {
+		tags = append(tags, fmt.Sprintf("resource-group:%s", metadata.ResourceGroup))
+	}
```
resource group is tagged as "resource_group"

graybjones commented on pkg/util/azure/azure_tags.go:
```diff
@@ -0,0 +1,50 @@
+	}
+	if metadata.Zone != "" {
+		tags = append(tags, fmt.Sprintf("zone:%s", metadata.Zone))
+	}
```
The crawler does not tag zone.

jipperinbham committed:
address initial PR feedback:

- return nil instead of empty slice for error conditions
- add check to ensure azure provider is enabled before collecting metadata

hush-hush commented:
Hi @jipperinbham,

Moving this to 7.23.0. Could you answer @graybjones comments ?

Thanks.
